---
title: "Small Cell Lung Cancer"
author:
- El Mehdi Baknine,  s194533
- Jakob Frostholm Højgaard, s194527
- Jonathan Dragestad Møller, s184243
- Mikkel Niklas Rasmussen, s193518
- Thomas Malthe Mølgaard Tams, s204540
date: '2023-11-28'
format:
  revealjs:
    theme: solarized
    logo: dtu_logo.png
    slide-number: TRUE
    embed-resources: true
---


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE)
library("tidyverse")
library("here")
```

## Introduction {.smaller}
### Paper and data source 
**Title:** Comprehensive genomic profiles of small cell lung cancer, George J. et. al. (2015)  



```{r out.width="80%",fig.align='center', margin= 0.1}
knitr::include_graphics(here("doc/data_chart_2.png"))
```


::: {style="text-align: center; margin-top: 1em"}
Purpose: Identify different small cell lung cancer profiles
:::


## Methods
### XXXXXXXXX

<br/>


::: {.fragment .fade-left}
Bullet point 1
:::

::: {.fragment .fade-left}
Bullet point 2
:::


::: {.fragment .fade-left}
Bullet point 3
:::


::: {.fragment .fade-left}
Bullet point 4
:::


::: {.fragment .fade-left}
Bullet point 5
:::

::: {.fragment .fade-left}
Bullet point 6
:::






## Data set overview


:::: {.columns}

::: {.column width="30%"}
::: {.fragment .fade-in}
**Loading**  

- Dimensions: 81, 31669 

- 30 metadata

- 31639 gene expression
:::

:::

::: {.column width="30%"}
::: {.fragment .fade-in}
**Cleaning**  

- New dimensions: 81, 31669

- Check for duplicates in SampleIDs

- Clean weird variables

- Check NAs 


:::
:::

::: {.column width="30%"}
::: {.fragment .fade-in}
**Augmenting**  

- New dimensions: 81, 31772



:::
:::

::::
## Overview of metadata 
:::: {.columns}

::: {.column width="50%"}

![](../results/04_ethnicity_dist.png){.absolute top="100" left="-200" width="600" height="600"}

:::

::: {.column width="50%"}
::: {.fragment .fade-in}
![](../results/04_patient_gender_count_age_dist.png){.absolute top="100" left="410" width="800" height="600"}

:::
:::
::::

## Overview of metadata 
:::: {.columns}

::: {.column width="50%"}
![](../results/04_smoking_status_dist.png){.absolute top="100" left="-200" width="600" height="600"}
:::

::: {.column width="50%"}
::: {.fragment .fade-in}
![](../results/04_smoking_history_survival_status.png){.absolute top="100" left="410" width="800" height="600"}
:::
:::
::::

## Results
:::: {.columns}

::: {.column width="50%"}

![](../results/04_treatment_ridgeplot.png){.absolute top="100" left="-200" width="800" height="800"}
:::

::: {.column width="50%"}
::: {.fragment .fade-in}
![](../results/06_heatmap_plot.png){.absolute top="100" left="610" width="700" height="750"}
:::
:::
::::
<!-- K-means  -->
<!-- Hierchichal clustring  -->
<!-- t-test : N0 : There is no significant difference between clusters. -->
<!-- To get low-high expression signature -->


## 
:::: {.columns}

::: {.column width="50%"}
![](../results/07_associated_genes_vulcano_plot.png){.absolute top="0" left="-200" width="700" height="800"}
:::

::: {.column width="50%"}
::: {.fragment .fade-in}
![](../results/07_associated_genes_CI.png){.absolute top="0" left="520" width="700" height="800"}
:::
:::
::::


## PCA 
![](../results/05_PCA_4_feature.png){.absolute top="100" left="100" width="800" height="700"}


## PCA2 
:::: {.columns}

::: {.column width="50%"}
![](../results/05_PCA_side_by_side_cluster_feature.png){.absolute top="100" left="-200" width="8    00" height="600"}
:::

::: {.column width="50%"}
::: {.fragment .fade-in}
![](../results/05_PCA_DET_boxplot_pc1_cluster.png){.absolute top="90" left="620" width="700" height="700"}
:::
:::
::::


## Executable Code

```{r}
#| echo: false
#| fig-width: 10
#| fig-height: 10
library(ggplot2)
ggplot(mtcars, aes(hp, mpg, color = am)) +
  geom_point() +
  geom_smooth(formula = y ~ x, method = "loess")
```


## Plot1 {chalkboard-buttons="true"}
```{r out.width="80%",fig.align='center'}
knitr::include_graphics(here("results/07_associated_genes_CI.png"))
```

## Some exploratory data analysis
```{r out.width="100%", fig.align='center'}
knitr::include_graphics(here("results/07_associated_genes_vulcano_plot.png"))
```




## Conclusion {.smaller}
