---
title: "01 - Loading data"
subtitle: "R for Bio Data Science"
format: html
author: Mikkel Niklas Rasmussen (s193518)
date: "2023-11-08"
editor: visual
---

## Load libraries

```{r}
library("tidyverse")
library("here")
library("readxl")
```

## Load data

```{r}
# Load metadata from sheet 1
meta_data <- read_excel(
  here("data/_raw/raw_data.xlsx"),
  sheet = "Suppl_Table1",
  skip = 3,
  na = c("NA", "--")
)

# Load RNA-seq expression data from sheet 10
expr_data <- read_excel(
  here("data/_raw/raw_data.xlsx"),
  sheet = "Suppl_Table10",
  skip = 1,
  na = "NA"
)
```

## Combine the metadata and expression sheets from the excel table

First we will need to transpose the RNA-seq gene expression dataframe before joining it with the meta data. We

```{r}
# Check if there are duplicated genes
expr_data |>
  group_by(gene) |>
  filter(n() > 1) |>
  summarise(n = n()) |>
  arrange(desc(n))
```
We see that there are a lot of duplicated genes. This is due to multiple transcript mapping to the the same gene. Here we will sum over the rows where multiple transcripts map to the same gene
```{r}
# Sum the FPKM values across multiple transcrips mapping
# to the same gene
expr_data_sum <- expr_data |>
  select(-refseq) |>
  group_by(gene) |>
  summarise(across(
                   where(is.numeric),
                   sum
                  )
  )

# We then need to transpose the expression dataframe
# so that is can be joined with the metadata
expr_data_sum_trans <- expr_data_sum |>
  pivot_longer(
    -gene,
    names_to = "sample_id"
  ) |>
  pivot_wider(
    names_from = gene, 
    values_from = value
  )

# Let's count how many samples have RNA-seq data
meta_data |>
  group_by(`RNA-seq`) |>
  count()

# Before joining we will check which samples will not be
# joined
meta_data |>
  anti_join(
    expr_data_sum_trans,
    by = join_by(`Sample-ID` == sample_id)
  ) |>
  filter(`RNA-seq` == "no") |>
  count()
```

We see that the samples that are not possible to join are for samples where the meta data also states that the is no RNA-seq data for the sample. Hence, it makes sense that it is not possible to join these.

```{r}
# Inner join the data
all_data <- meta_data |>
  inner_join(
    expr_data_sum_trans, 
    by = join_by(`Sample-ID` == sample_id)
  )
```

## Save the combined data

```{r}
# Save in data directory, as .tsv file
write_tsv(
  all_data, 
  here("data/01_dat_load.tsv")
)
```
